node/src/crypto/hash.rs

```diff
@@ -172,6 +174,92 @@ impl From<Blake2bHash> for Digest {
     }
 }
 
+/// Sentinel hash to be used for hashing options in the case of [None].
+pub const SENTINEL0: Digest = Digest([0u8; 32]);
+/// Sentinel hash to be used by [hash_slice_rfold]. Terminates the fold.
+pub const SENTINEL1: Digest = Digest([1u8; 32]);
+/// Sentinel hash to be used by [hash_vec_merkle_tree] in the case of [
+pub const SENTINEL2: Digest = Digest([2u8; 32]);
+
+/// Hashes a pair of [Digest]s.
+pub fn hash_pair(hash1: &Digest, hash2: &Digest) -> Digest {
+    let mut to_hash = [0; Digest::LENGTH * 2];
+    to_hash[..Digest::LENGTH].copy_from_slice(&(hash1.to_array())[..]);
+    to_hash[Digest::LENGTH..].copy_from_slice(&(hash2.to_array())[..]);
+    hash(&to_hash)
+}
+
+/// Hashes a [Vec<Digest>] into a single [Digest] by constructing a [Merkle tree][1].
+/// Reduces pairs of elements in the [Vec<Digest>] by repeatedly calling [hash_pair].
+/// This hash procedure is suited to hashing [BTree]s.
+///
+/// The pattern of hashing is as follows.  It is akin to [graph reduction][2]:
+///
+/// ```text
+/// a b c d e f
+/// |/  |/  |/
+/// g   h   i
+/// | /   /
+/// |/   /
+/// j   k
+/// | /
+/// |/
+/// l
+/// ```
+///
+/// Returns the empty [Digest] when the input is empty.
+///
+/// [1]: https://en.wikipedia.org/wiki/Merkle_tree
+/// [2]: https://en.wikipedia.org/wiki/Graph_reduction
+pub fn hash_vec_merkle_tree(vec: Vec<Digest>) -> Digest {
+    vec.into_iter()
+        .tree_fold1(|x, y| hash_pair(&x, &y))
+        .unwrap_or(SENTINEL2)
+}
+
+/// Hashes a [BTreeMap].
+pub fn hash_btree_map<K, V>(btree_map: &BTreeMap<K, V>) -> Result<Digest, bytesrepr::Error>
+where
+    K: ToBytes,
+    V: ToBytes,
+{
+    let mut kv_hashes: Vec<Digest> = Vec::with_capacity(btree_map.len());
+    for (key, value) in btree_map.iter() {
+        kv_hashes.push(hash_pair(&hash(key.to_bytes()?), &hash(value.to_bytes()?)))
+    }
+    Ok(hash_vec_merkle_tree(kv_hashes))
+}
+
+/// Hashes a `&[Digest]` using a [right fold][1].
+///
+/// This pattern of hashing is as follows:
+///
+/// ```text
+/// hash_pair(a, &hash_pair(b, &hash_pair(c, &SENTINEL)))
+/// ```
+///
+/// Unlike Merkle trees, this is suited to hashing heterogeneous lists we may wish to extend in the
+/// future (ie, hashes of data structures that may undergo revision).
+///
+/// Returns [SENTINEL1] when given an empty [Vec] as input.
+///
+/// [1]: https://en.wikipedia.org/wiki/Fold_(higher-order_function)#Linear_folds
+pub fn hash_slice_rfold(slice: &[Digest]) -> Digest {
+    slice
+        .iter()
+        .rfold(SENTINEL1, |prev, next| hash_pair(next, &prev))
+}
+
+/// Hashes a `&[Digest]` using a [right fold][1]. Uses `proof` as a Merkle proof for the missing
+/// tail of the slice.
+///
+/// [1]: https://en.wikipedia.org/wiki/Fold_(higher-order_function)#Linear_folds
+pub fn hash_slice_with_proof(slice: &[Digest], proof: Digest) -> Digest {
+    slice
+        .iter()
+        .rfold(proof, |prev, next| hash_pair(next, &prev))
+}
```

Comment:
I'm wondering whether we shouldn't change the right fold to the left fold here (and in the vec/tree as well). The right fold has a property that new elements are appended (vs prepending in the left fold) which could make it more performant if we ever need to add new elements into it. I'm thinking about a vector of elements (digests) that we extend – prepending would most probably require allocating and rewriting all content of a vector into new one, while appending would only extend the size of the vector without rewriting the content.

Reply:
I think the idea was that once you know that the era report is the _second_ field, you'll always be able to check proofs of the era report based on the hash, even if other fields get _appended_ in a later protocol version. It's true that appending changes all those hashes, but the length of the era report proof will always be 2. For that you need an `rfold`.